thonny/running.py

```diff
@@ -916,8 +974,42 @@ def _get_environment(self):
             env["THONNY_DEBUG"] = "1"
         elif "THONNY_DEBUG" in env:
             del env["THONNY_DEBUG"]
+
+        add_env_file = self._load_env_file()
+
+        chained_env = collections.ChainMap(add_env_file, env)
+        chained_env = dict(chained_env)
+
+        # un-set keys
+        env = EnvFile()
+        env.set_env(chained_env)
+        env.remove_unused()
+
+        return env.get_env()
+
+    def _load_env_file(self):
+        # todo: configuration menu -> Tools > Options > Run & Debug
+        env_file = os.environ.get("THONNY_ENV_FILE", ".env")
+        script_dir = self._get_launcher_python_script_dir()
+
+        env_file = os.path.join(script_dir, env_file)
+
+        if os.path.exists(env_file) and os.path.isfile(env_file):
+            env = EnvFile(env_file)
+            env.load()
+            if env.has_errors():
+                logger.error("env-file errors in lines: " + str(env.err_lines))
+                if env.excep:
+                    logger.exception("env-file" + str(env.excep))
+            env = env.get_env()
+        else:
+            env = {}
+
         return env
 
+    def _get_launcher_python_script_dir(self):
+        return self._get_launcher_with_args()[-1]
```

Comment:
Here you are simply returning last argument -- this doesn't look right.

I would look for the `.env` from the working directory of the backend process. See the first option in "Tools => Options => Run & Debug"

Reply:
i got this by putting a print statement into [running.py -> _start_background_process](https://github.com/thonny/thonny/blob/6e7b9cf572a46cd95bdd38a5411064b9649d8b7f/thonny/running.py#L1013)
and checked what happens when i press "run" button in thonny.

but since i was also quite happy with my findings i put it into an own function because i think that sort of function should be available e.g. `get_workbench().get_launcher_python_script_dir()`

Reply:
I see, I had forgotten, that _get_launcher_with_args puts working directory as the last argument.

IMO it would be more direct to use `self.get_cwd()`.

Also, I now noticed that you added stuff to the SubprocessProxy, but I think it should be in a more specific class -- CPythonProxy. SubprocessProxy is used also with MicroPython and Ssh backends, and the .env file is not relevant there (at least not directly). For example, if you select "Remote Python 3" backend, then you probably want the .env file to be applied only on the remote host.

Reply:
i rebased my branch to master.
regarding SubprocessProxy i will have to do more investigations... 

one maybe stupid question here:
why not having sort of propagation of env settings here to MicroPython and Ssh backends?
so that it is also available there like you mentioned already... 

> For example, if you select "Remote Python 3" backend, then you probably want the .env file to be applied only on the remote host.

Reply:
> why not having sort of propagation of env settings here to MicroPython and Ssh backends?
> so that it is also available there like you mentioned already

With SSH back-end it wouldn't propagate anyway over SSH connection -- the variables need to be set explicitly on remote host (https://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.exec_command). And it wouldn't be harmless either -- in principle a variable that is fit for the remote process may mess up the local process.

Reply:
The local env is not touched by this implementation.
Or? 

So with the given link to exec_command documentation the propagation should be easy to implement.

How do we deal with this now?
I m asking because I can not do much more this year, and I m somehow unaware of thonny 4.0 release timeline...

So one possibility could be merge current code and open a new issue regarding the SSH limitation and propagation Todo.

So... ???

Reply:
> The local env is not touched by this implementation.

Currently theres two local processes when you're using ssh backend -- the UI process and another local process, which communicates with the remote host. I'm may merge the two for 4.0, though.

I hope to get the first beta of 4.0 out before the end of December. If you can't proceed by then, then I guess I'll merge the current state and fix it as I see fit.

Reply:
Probably I have some time between Xmas and new year. 
Otherwise just merge and create an issue, or add the missing SSH integration code